libbeat/common/transport/tlscommon/tls_config.go

```diff
@@ -116,3 +121,82 @@ func (c *TLSConfig) BuildModuleConfig(host string) *tls.Config {
 	config.ServerName = host
 	return config
 }
+
+// MakeVerifyPeerCertificate creates the verification combination of checking certificate pins and skipping host name validation depending on the config
+func MakeVerifyPeerCertificate(cfg *TLSConfig) verifyPeerCertFunc {
+	pin := len(cfg.CASha256) > 0
+	skipHostName := cfg.Verification == VerifyCertificate
+
+	if pin && !skipHostName {
+		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			return VerifyCAPin(cfg.CASha256, verifiedChains)
+		}
+	}
+
+	if pin && skipHostName {
+		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			_, _, err := VerifyCertificateExceptServerName(rawCerts, cfg)
+			if err != nil {
+				return err
+			}
+			return VerifyCAPin(cfg.CASha256, verifiedChains)
+		}
+	}
+
+	if !pin && skipHostName {
+		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			_, _, err := VerifyCertificateExceptServerName(rawCerts, cfg)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// VerifyCertificateExceptServerName is a TLS Certificate verification utility method that verifies that the provided
+// certificate chain is valid and is signed by one of the root CAs in the provided tls.Config. It is intended to be
+// as similar as possible to the default verify (go/src/crypto/tls/handshake_client.go:259), but does not verify
+// that the provided certificate matches the ServerName in the tls.Config.
+func VerifyCertificateExceptServerName(
+	rawCerts [][]byte,
+	c *TLSConfig,
+) ([]*x509.Certificate, [][]*x509.Certificate, error) {
+	// this is where we're a bit suboptimal, as we have to re-parse the certificates that have been presented
+	// during the handshake.
+	// the verification code here is taken from crypto/tls/handshake_client.go:259
+	certs := make([]*x509.Certificate, len(rawCerts))
+	for i, asn1Data := range rawCerts {
+		cert, err := x509.ParseCertificate(asn1Data)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "tls: failed to parse certificate from server")
+		}
+		certs[i] = cert
+	}
+
+	var t time.Time
+	if c.Time != nil {
+		t = c.Time()
+	} else {
+		t = time.Now()
+	}
+
+	// DNSName omitted in VerifyOptions in order to skip ServerName verification
+	opts := x509.VerifyOptions{
+		Roots:         c.RootCAs,
+		CurrentTime:   t,
+		Intermediates: x509.NewCertPool(),
+	}
+
+	for i, cert := range certs {
+		if i == 0 {
+			continue
+		}
+		opts.Intermediates.AddCert(cert)
+	}
```

Comment:
Ah good catch, this part was a copy and paste job from our repo I should have caught. And it looks like upstream has it this way too so idk why we did it all silly https://github.com/golang/go/blob/master/src/crypto/tls/handshake_client.go#L842

Reply:
in case it is copied from the golang/go code base we need to update the file header licence info as well. maybe we can move it to a separate verify.go file?

Reply:
For this i just copied it from the ECK repo, which should be okay since we can relicense code we wrote without issue. That's a good catch though and I'm not sure of its provenance. It looks like we added it here https://github.com/elastic/cloud-on-k8s/pull/929 but unfortunately the author is out

Reply:
Looking at it more closely, the requirements in the golang license are just a copyright notice, which is already included since this package imports crypto/tls, so I think even if we copied the code line by line (which I do not expect is the case) we are good as is? I am not a lawyer though.

Reply:
The go repo is BSD licensed. I think BSD requires copies to include a copyright notice.

Reply:
@urso it looks like having the golang header in `verify.go` causes the license header CI check to fail. It's not clear to me that the header is required. From the golang src:

>//    * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.

which I believe is covered by including it in the NOTICE.txt at the root of the repo.